onnx/defs/tensor_proto_util.cc

```diff
@@ -31,15 +31,19 @@ namespace ONNX_NAMESPACE {
     return t;                                                   \
   }
 
-#define DEFINE_PARSE_DATA(type, typed_data_fetch)                              \
+#define DEFINE_PARSE_DATA(type, typed_data_fetch, tensorproto_datatype)        \
   template <>                                                                  \
   const std::vector<type> ParseData(const TensorProto* tensor_proto) {         \
-    std::vector<type> res;                                                     \
     if (!tensor_proto->has_data_type() ||                                      \
       tensor_proto->data_type() == TensorProto_DataType_UNDEFINED) {           \
       fail_shape_inference("The type of Tensor: ", tensor_proto->name(),       \
         " is undefined so it cannot be parsed.");                              \
-    } else if (tensor_proto->has_data_location() &&                            \
+    } else if (tensor_proto->data_type() != tensorproto_datatype) {            \
```

Comment:
do you need the if check above? Can you combine if and else if into 1 if loop?
You can update the error message here to say something like 
```
fail_shape_inference("ParseData type mismatch for tensor: ", tensor_proto->name(), "Expected", <>, "Actual", <>);
```

Reply:
I think these two if are required since TensorProto_DataType_UNDEFINED is not in the existing type-string map? Two different error messages look clear to me for two different kinds of errors. Let me know if you still think they should be merged since we can still extend the map with TensorProto_DataType_UNDEFINED. ("undefined" to TensorProto_DataType_UNDEFINED)

Reply:
No need to extend the map... as ToDataTypeString will throw an error if type is not in its mapping... Maybe it is OK to keep these 2 checks... Can you make sure we have enough tests for ParseData

Reply:
Added one test to test invalid type for ParseData: strict shape inference should catch it and normal shape inference should not produce unexpected shape

Reply:
Added another test to test undefined Tensor type